File: csmal/csmal/Step1ReadPrint.cs
using System;
using Deveel;

namespace csmal
{
    public class Step1ReadPrint
    {
        private static void Main(string[] args)
        {
            while (true)
            {
                var line = Readline.ReadLine("REPL> ");
                if (line == null)
                {
                    break;
                }
                Console.WriteLine(Rep(line));
            }
        }

        private static MalType Read(string arg)
        {
            return Reader.ReadStr(arg);
        }

        private static MalType Eval(MalType arg)
        {
            return arg;
        }

        private static string Print(MalType arg)
        {
            return Printer.PrStr(arg);
        }

        public static string Rep(string arg)
        {
            try
            {
                return Print(Eval(Read(arg)));
            }
            catch (NoTokensException)
            {
                return ""; // in this case, by design, we don't worry about the input
            }
        }
    }
}

File: gblrust/src/bin/step1_read_print.rs
extern crate mal;
extern crate rustyline;
use mal::types::MalType;
use mal::reader;
use rustyline::Editor;

fn read(a: &str) -> Result<MalType, String> {
    reader::read_str(a)
}

// EVAL continues to simply return its input but the type is now a mal data type.
fn eval(x: MalType) -> MalType {    
    x
}

fn print(x: MalType) -> String {
    // and the PRINT function to call printer.pr_str.
    x.pr_str()
}

fn rep(a: &str) -> Result<String, String> {
    match read(a) {
        Ok(x) => Ok(print(eval(x))),
        Err(y) => Err(y)
    }
}

fn main() {
    let mut rl = Editor::<()>::new();
    loop {
        let readline = rl.readline("user> ");
        let input = match readline {
            Ok(input) => input,
            Err(y) => {
                println!();
                break;
            },
        };
        
        match rep(&(input.to_owned()[..])) {
            Ok(x) => println!("{}", x),
            Err(y) => println!("{}", y)
        }
    }
}
